policy/policy.go

```diff
@@ -44,85 +45,198 @@ func (policy *PolicySvc) Routes() common.Routes {
-	topoUrl, err := client.GetServiceUrl("topology")
+	if segmentIDToUse != "" {
+		tenantsUrl := fmt.Sprintf("%s/tenants/%s/segment/%s", tenantSvcUrl, ten.ID, segmentIDToUse)
+		segment := &tenant.Segment{}
+		err = policy.client.Get(tenantsUrl, &segment)
+		if err != nil {
+			return err
+		}
+		endpoint.SegmentNetworkID = segment.Seq
+	}
+	return nil
+}
+
+// augmentPolicy augments the provided policy with information gotten from
+// various services.
+func (policy *PolicySvc) augmentPolicy(policyDoc *common.Policy) error {
+	// Get info from topology service
+	topoUrl, err := policy.client.GetServiceUrl("topology")
 	if err != nil {
-		return nil, err
+		return err
 	}
 
+	// Query topology for data center information
+	// TODO move this to root
 	index := common.IndexResponse{}
-	err = client.Get(topoUrl, &index)
+	err = policy.client.Get(topoUrl, &index)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	tenantSvcUrl, err := client.GetServiceUrl("tenant")
+	dcURL := index.Links.FindByRel("datacenter")
+	dc := common.Datacenter{}
+	err = policy.client.Get(dcURL, &dc)
 	if err != nil {
-		return nil, err
+		return err
+	}
+	log.Printf("Policy server received datacenter information from topology service: %v\n", dc)
+	policyDoc.Datacenter = dc
+
+	for _, endpoint := range policyDoc.AppliedTo {
+		err = policy.augmentEndpoint(&endpoint)
+		if err != nil {
+			return err
+		}
 	}
 
-//	for _, endpoint := range policyDoc.AppliedTo {
-//	}
-	tenantsUrl := fmt.Sprintf("%s/tenants", tenantSvcUrl)
-	var tenants []tenant.Tenant
-	err = client.Get(tenantsUrl, &tenants)
+	for _, endpoint := range policyDoc.Peers {
+		err = policy.augmentEndpoint(&endpoint)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// distributePolicy distributes policy to all agents.
+// TODO how should error handling work here really?
+func (policy *PolicySvc) distributePolicy(policyDoc *common.Policy) error {
+	hosts, err := policy.client.ListHosts()
 	if err != nil {
-		return nil, err
+		return err
 	}
-	found := false
-	var i int
-	tenantName := "TODO"
-	for i = range tenants {
-		if tenants[i].Name == tenantName {
-			found = true
-			break
+	errStr := make([]string, 0)
+	for _, host := range hosts {
+		// TODO make schema configurable
+		url := fmt.Sprintf("http://%s:%d/policies", host.Ip, host.AgentPort)
+		result := make(map[string]interface{})
+		err = policy.client.Post(url, policyDoc, result)
+		log.Printf("Agent at %s returned %v", host.Ip, result)
+		if err != nil {
+			errStr = append(errStr, fmt.Sprintf("Error applying policy %d to host %s: %v. ", policyDoc.ID, host.Ip, err))
 		}
 	}
-	if !found {
-		return nil, errors.New("Tenant with name " + tenantName + " not found")
+	if len(errStr) > 0 {
+		return common.NewError500(errStr)
 	}
+	return nil
+}
 
-	segmentsUrl := fmt.Sprintf("/tenants/%s/segments", "")
-	var segments []tenant.Segment
-	err = client.Get(segmentsUrl, &segments)
+// deletePolicy deletes policy...
```

Comment:
Additional docstring: The policy-ID is specified as a path variable and retrieved through the RestContext.